llvm/lib/SYCLLowerIR/SYCLOptimizeBarriers.cpp

```diff
@@ -0,0 +1,704 @@
+            static_cast<uint32_t>(MemorySemantics::WorkgroupMemory);
+        B.CI->setArgOperand(2, ConstantInt::get(Int32Ty, NewSem));
+        B.Semantic = NewSem;
+      }
+      B.CI->setArgOperand(1, ConstantInt::get(Int32Ty, static_cast<uint32_t>(
+                                                           Scope::Workgroup)));
+      B.MemScope = Scope::Workgroup;
+      Changed = true;
+    }
+  }
+
+  return Changed;
+}
+
+// True if BD is the first real instruction of the function.
+static bool isAtKernelEntry(const BarrierDesc &BD) {
+  BasicBlock &Entry = BD.CI->getFunction()->getEntryBlock();
+  if (BD.CI->getParent() != &Entry)
+    return false;
+
+  for (Instruction &I : Entry) {
+    if (&I == BD.CI)
+      break;
+    if (classifyMemScope(&I) != RegionMemScope::None)
+      return false;
+  }
+
+  return true;
+}
+
+// True if BD is immediately before a return/unreachable and nothing follows.
+static bool isAtKernelExit(const BarrierDesc &BD) {
+  BasicBlock *BB = BD.CI->getParent();
+  Instruction *Term = BB->getTerminator();
+  if (!isa<ReturnInst>(Term) && !isa<UnreachableInst>(Term))
+    return false;
+
+  for (Instruction *I = BD.CI->getNextNode(); I && I != Term;
+       I = I->getNextNode())
+    if (classifyMemScope(I) != RegionMemScope::None)
+      return false;
+
+  return BD.CI->getNextNonDebugInstruction() == Term;
+}
+
+// Remove barriers that appear at the very beginning or end of a kernel
+// function.
+static bool
+eliminateBoundaryBarriers(SmallVectorImpl<BarrierDesc *> &Barriers) {
+  bool Changed = false;
+  for (auto *BPtr : Barriers) {
+    BarrierDesc &B = *BPtr;
+    if (!B.CI)
+      continue;
+    // FIXME?: do we _really_ need this restriction? If yes - should it be
+    // applied for other transformations done by the pass?
+    if (B.CI->getFunction()->getCallingConv() != CallingConv::SPIR_KERNEL)
+      continue;
+    if (isAtKernelEntry(B) || isAtKernelExit(B))
+      Changed |= eraseBarrierWithITT(B);
+  }
+  return Changed;
+}
+
+} // namespace
+
+PreservedAnalyses SYCLOptimizeBarriersPass::run(Function &F,
+                                                FunctionAnalysisManager &AM) {
+  SmallVector<BarrierDesc, 16> Barriers;
+  BBMemInfoMap BBMemInfo;
+  BarriersMap BarriersByBB;
+  SmallVector<BarrierDesc *, 16> BarrierPtrs;
+
+  // Analyse the function gathering barrier and memory scope of the region info.
+  collectBarriersAndMemInfo(F, Barriers, BBMemInfo);
+  for (auto &B : Barriers)
+    BarriersByBB[B.CI->getParent()].push_back(B);
+
+  for (auto &Pair : BarriersByBB)
+    for (auto &BD : Pair.second)
+      BarrierPtrs.push_back(&BD);
+
+  bool Changed = false;
+  // First remove 'at entry' and 'at exit' barriers if the fence nothing.
+  Changed |= eliminateBoundaryBarriers(BarrierPtrs);
+  // Then remove redundant barriers within a single basic block.
+  for (auto &BarrierBBPair : BarriersByBB)
+    Changed = eliminateBackToBackInBB(BarrierBBPair.first, BarrierBBPair.second,
```

Comment:
can `eliminateBackToBackInBB` be merged into `eliminateDominatedBarriers`? `eliminateBackToBackInBB` is just a special case of the latter in that all barriers are in a single BB, right?